Implement build system using self-hosted aarch64 runners, GitHub `needs` jobs feature and reusable workflows

mathbunnyru opened the pull request:
I've decided to implement our build system from scratch.

Ideas, that came to my mind:
- one job in a workflow only does one essential job, only one platform
- so, we will have lots of jobs, like `build-x86-minimal-notebook`, but we can easily write dependencies between them
- share everything as GitHub artifacts between these jobs
- amd64 and aarch64 are almost independent
- do not even try to merge tags between amd64 and aarch64, let's add the `aarch64-` prefix for all arm builds

Implementation details:
- I rely heavily on reusable workflows. This way there is almost zero code duplication
- The duplication is only needed in the main `docker.yml` workflow. This file can be seen as a simple config file of build dependencies
- No build system is actually needed because we will be relying on the `needs` feature of GitHub
- Heavily rely on well-made GitHub workflows `actions/upload-artifact` and `actions/download-artifact` to pass the image between jobs

TODO: 
- [x] aarch64: self-hosted runners
- [x] self-hosted runners: easy setup docs
- [x] adapt taggers to support aarch64 prefix
- [x] adapt manifest creation for the new build system
- [x] delete `buildx` Makefile parts (or event the Makefile itself)
- [x] add all images to the main workflow
- [x] fix documentation where it mentions `aarch64` or `make` commands
- [x] test, that only fresh images are used - add an empty file to the image and check, that it persists in every image built

Fix: https://github.com/jupyter/docker-stacks/issues/1407
Fix: https://github.com/jupyter/docker-stacks/issues/1530
Fix: https://github.com/jupyter/docker-stacks/issues/1402
Fix: https://github.com/jupyter/docker-stacks/issues/1401
Fix: https://github.com/jupyter/docker-stacks/issues/1203
Supersedes: https://github.com/jupyter/docker-stacks/pull/1631
Fix is no longer needed for: https://github.com/jupyter/docker-stacks/issues/1539

Upsides:
- (all the issues above)
- no need to use Makefile
- no `sudo rm -rf` to maximise space, because we will have a small amount of images in the local cache (hopefully)
- the summary of jobs shows a nice dependency graph
- aarch64 are first-class images now, there are no special hacks for building them (except for native runners and sharing state in VM)
- adding new platforms is theoretically possible now, but we will have to have native runners for them as well (in practice - do not expect new platforms)
- no need for a separate amd64 workflow
- easily adaptable for GitHub native aarch64 runners when they will be available
- no need to manually create sections to make GitHub steps look good - I think it will be easier to jump right into the error if one occurs
- QEMU aarch64 bug fix removal (no longer needed), which will allow faster `mamba` execution on multi-CPU machines

Downsides:
- Actions in self-hosted runners are run in VMs environment: this means worse security and some work to clean up the environment.
- in theory should be extremely fast, because everything that can be run in parallel, will be. The overhead of using a clean environment is small. But `docker save/load` and `uploading/downloading` artifacts is quite slow.
- maintenance self-hosted GitHub runners
- we will have an `aarch64-` prefix for every arm tag
- local experience will be a bit different from the remote one, I don't want to support `docker buildx` and multi-plaform in Makefile
- to add a new core image, more lines of changes will be required. But since we haven't added new images for many years, this is not a problem.

mathbunnyru commented:
This is just a beginning.

<img width="919" alt="Снимок экрана 2022-05-12 в 16 42 16" src="https://user-images.githubusercontent.com/12270691/168089326-6952d476-c154-4811-af9d-8bd366228e04.png">

mathbunnyru commented:
Current limitation: https://github.com/docker/buildx/issues/847

mathbunnyru commented:
@trallard @consideRatio @parente @romainx @maresb this is ready for review.

The only thing it doesn't support is `aarch64`.

The only thing needed to add this support is one of two things:
- Being able to `docker buildx build`, then upload the file, then download it on another job (machine) and build the imaged bases on the pulled one (https://github.com/docker/buildx/issues/847)
- Having self-hosted aarch64 runners

I prefer the second option, as it seems to be much more faster and doesn't create many hacks for this project.
But, I don't have any experience with running my own GitHub runner.
I will give it a try, when I have some inspiration :)

Please, share your ideas about this approach.
I will also publish total build time when the build finishes.

mathbunnyru commented:
Finally, this workflow runs in 50 minutes.
As we add aarch64 images, this time will not increase (assuming that building aarch64 images takes similar time to build).

Running in the master branch takes some time to push the images, but this time there will be no rebuilds during the push, so this should be OK.

mathbunnyru commented:
I would also like to personally thank @trallard.
I have learned a few tricks from your PR, like using `needs` for jobs and using docker save/load + actions upload/download.

maresb commented:
Hey @mathbunnyru this looks very exciting! I'm on the road and sadly don't have time (or likely ability) to help properly review this, but there is one thing which seems strange to me, where if I'm very lucky I might be able to say something helpful. (Otherwise, please forgive the noise.)

>  Being able to docker buildx build, then upload the file, then download it on another job (machine) and build the image based on the pulled one

What is "the file" in this context? Is it a Docker image or an actual file?

In case you want to push/pull images which should be used as cache, then it is possible to coax buildx into doing this, given that you set the (IMHO highly counterintuitive) correct environment variables and CLI flags. If this is the case, I might be able to help you somewhat by comparing with a few things I already have up-and-running.

Otherwise, this looks more elaborate than anything I've ever done with GH Actions, so I'll probably use it as a reference in some future project, so thanks in advance! :smile:

trallard commented:
I can have a look and review this on Monday/Tuesday 🙏🏽  sorry for the silence recently work has been busy

mathbunnyru commented:
AFAIK, using GitHub container registry in branches is not possible 😢 
You're right, these images are used for several purposes:
- for a subsequent image, as you told
- to run this image and run tests in it / create manifests

maresb commented:
I have a recent project using buildx and docker-compose on GitLab, so it's not something I can directly give you, but the essential command seems to be:

```bash
export DOCKER_BUILDKIT=1
docker buildx build --build-arg BUILDKIT_INLINE_CACHE=1 --cache-from from-image ...
```

Setting `BUILDKIT_INLINE_CACHE=1` as a build arg will make the resulting image usable as cache when pulled.

The `--cache-from` argument indicates which pulled images are allowed to be used as cache.

As long as I'm not missing anything, this should allow you to reuse images between different runners and such.

That's a shame about GHCR. Perhaps you could set up a separate DockerHub account? All you need is some generic registry.

mathbunnyru commented:
Thanks, I will try it out!
And could you tell me, how exactly do you pass this cache between jobs?
Are you using `docker save` or are you using the registry?

> That's a shame about GHCR. Perhaps you could set up a separate DockerHub account? All you need is some generic registry.

I don't think this would work.
I mean, to be able to run in branches, I will have to add the credentials, not as secrets, but to write them in code.
Our images then will not be safe, unfortunately.

I thought about using services in GitHub and run the registry as a service, but I have zero experience with that. And especially, using reuseable workflows, which is a pretty new feature of GitHub.

maresb commented:
> And could you tell me, how exactly do you pass this cache between jobs?

Using `docker push` and `docker pull`, with the registry as global storage. Does this make sense, at least in theory?

I'm a bit confused about how you'd like to protect the images. (Sorry I don't yet have an understanding of how secret sharing works on GitHub with multiple branches, perhaps this is why I'm confused?)

Ah, maybe I understand now... you effectively want a private registry for each branch? Would it be sufficient to use a single set of credentials, passed as secrets, but build images which have branch-specific tags?

Unfortunately I don't know much about GitHub in particular, so if GitHub isn't populating the credentials in your desired branches, I don't know how to help.

consideRatio commented:
> First of all, you can't use GitHub secrets in any branch other than the main branch.

I think there may be an exception for container registry access actually. At least you can make PRs read "private" registries, so maybe PRs can also push to them for specific images? To trial this, we need permissions in the jupter org itself, adding images to ghcr.io is not a project level thing but a org level thing.

I'm on vacation, will be back on monday, i can see if someone can help us trial this.

trallard commented on .github/workflows/docker.yml:
```diff
@@ -7,6 +7,14 @@ on:
   pull_request:
     paths:
```
Maybe for the less familiar with Actions add a comment at the top like:

"Composable and reusable GitHub action workflows" or something similar + the link to the docs section

mathbunnyru replied:
Sure, added this info both for reuseable workflows and for composite actions.

trallard commented on .github/workflows/docker.yml:
```diff
@@ -39,84 +54,181 @@ on:
 
+# https://docs.github.com/en/actions/using-jobs/using-concurrency
 concurrency:
-  # only cancel in-progress jobs or runs for the current workflow - matches against branch & tags
```
can we keep the comment? - it helps folks to understand what is being done here if they are less familiar with GH actions

trallard commented on docs/contributing/tests.md:
```diff
@@ -5,11 +5,7 @@ We greatly appreciate pull requests that extend the automated tests that vet the
-1. `make build-all-multi` - which builds all the Docker images
-2. `make test-all` - which tests the newly created Docker images
-   This `make` command builds and then tests every docker image.
+runs tests against pull requests submitted to the `jupyter/docker-stacks` repository:
```
```suggestion
runs tests against pull requests submitted to the `jupyter/docker-stacks` repository.
```

trallard commented on .github/workflows/docker-build-upload.yml:
```diff
@@ -0,0 +1,55 @@
+      - name: Create dev environment 📦
+        uses: ./.github/actions/create-dev-env
+
+      - name: Load parent built image to docker 📥
```
```suggestion
      - name: Load parent built image to Docker 📥
```

trallard commented on .github/workflows/docker-build-upload.yml:
```diff
@@ -0,0 +1,55 @@
+          # Full logs for CI build
+          BUILDKIT_PROGRESS: plain
+        shell: bash
+      - name: Save image as a file 💾
```
To avoid confusion
```suggestion
      - name: Save image as a tar for later use 💾
```

trallard commented on .github/workflows/docker-build-upload.yml:
```diff
@@ -0,0 +1,55 @@
+        run: docker save jupyter/${{ inputs.image }} -o /tmp/${{ inputs.image }}-${{ inputs.platform }}.tar
+        shell: bash
+
+      - name: Upload artifact 💾
```
```suggestion
      - name: Upload image as artifact 💾
```

trallard commented on .github/workflows/docker-tag-manifest-push.yml:
```diff
@@ -0,0 +1,69 @@
+name: Download image artifact from GitHub artifacts, tag it and push to DockerHub
```
```suggestion
name: Download Docker image from GitHub artifacts, tag and push it to DockerHub
```

trallard commented on .github/actions/load-image/action.yml:
```diff
@@ -0,0 +1,29 @@
+name: "Load image"
```
Keeping consistency 

```suggestion
name: Load Docker image
```

trallard commented on .github/workflows/docker-tag-manifest-push.yml:
```diff
@@ -0,0 +1,69 @@
+      - name: Create dev environment 📦
+        uses: ./.github/actions/create-dev-env
+
+      - name: Load image to docker 📥
```
```suggestion
      - name: Load image to Docker 📥
```

trallard commented on .github/workflows/docker.yml:
```diff
@@ -39,84 +54,181 @@ on:
+      platform: amd64
+      runsOn: ubuntu-latest
+
+  test-amd64-base:
```
Actually in hindsight why is the build and test in separate jobs - so that you can kick - off building the child image as soon as the parent is built?

An alternative to avoid the save + load time accumulation might be to test right after build - then save it as artefact for use in other jobs. 
This might delay some downstream builds but might shave some time on the load task

mathbunnyru replied:
Yes, for example we now have a big dependency chain (which is probably the longest one):
base->minimal->scipy->pyspark->all-spark.

If we make testing inside the build step, then we will make this chain run for 5min49s longer (we will be running images).
At the same time the longest run of tests including downloading image is 5min44s.

So, there is practically no difference.

At the same time, if we add a general test, which runs X seconds, in my architecture it will add around X seconds to complete, but when we run inside the build step, it will add around 5X time.
So, current solution is better adapted for future test suite increase.

The times are based on this particular run (I hope I haven't made a mistake):
https://github.com/jupyter/docker-stacks/actions/runs/2321552950


Also, I think, having tests in the same step as a build is OK, and reduces the github graph size, so I might try it after we merge (or reject) this PR.

trallard reviewed (commented):
Made some minor changes 

Added some comments on the logic here and there but I think this is a great improvement

trallard commented:
Another question - how does this then work with #1631 - I suppose a later PR once I manage to wrap this would be to add or integrate doit with all the workflows and build task to have one single way to build things moving forward

mathbunnyru commented:
> Before I forget - @mathbunnyru I love this work and the clarity gained from the parallel jobs is great ✨

Thank you ❤️

> I said I would help with the request - work has been a nightmare but will work on something and tag you all for input/reviews
> This will allow us to have the self-hosted runners - it might be a fun learning experience

Great news! I haven't started working on this actively, so I will wait for you 😂 

> As for using GHCR - I wonder if this will actually make the whole process any faster - ultimately we'd be pushing and pulling multiple images, multiple times - might need to do a comparison on saving as artifact + load time to make a decision

Agreed. I would like to hear from @consideRatio about his experience using GHCR in PRs and branches.

> Another question - how does this then work with #1631 - I suppose a later PR once I manage to wrap this would be to add or integrate doit with all the workflows and build task to have one single way to build things moving forward

To be honest, I think using doit might become an overkill.
I mean, it's a great tool to deal with dependencies.
But if this is merged, then Makefile is basically a transofrmation from `make build/base` to `docker build args...`, there is not much logic there.

@trallard thank you for your review.
I hope I haven't forgotten to answer any of your questions :)

mathbunnyru commented on .github/actions/create-dev-env/action.yml:
```diff
@@ -0,0 +1,16 @@
+name: "Build environment"
```
```suggestion
name: Build environment
```

mathbunnyru commented:
Small detail - even though we use a lot of steps, I see the consistent result in the build times.
It's around 50m each time.
Last time it was 47m 14s.
And the chain which takes the longest to run and is the actually the total time looks like this:
`build-base`->`build-minimal`->`build-base`->`build-scipy`->`build-pyspark`->`build-all-spark`->`test-all-spark`->`tag-push-tensorflow`(these jobs wait for images to be tested)->`wiki-update`.

trallard commented:
I do not know why I did not think of this when made my review 

But another thing to consider is we generally want to shift left (i.e test as early as possible) so following this I'd prefer having the test within the same job as the build. 
Why?
Imagine the case where we build `base` then continue building images before testing. If we defer the testing and this fails then we will have built images that will more than likely fail the tests too. 
So ideally we want to test and fail as early as possible - especially since the overall time seems unaffected

mathbunnyru commented:
Ok, let's try it out :)

mathbunnyru commented:
Used push force just a little bit, to be able to easily revert the commit.

mathbunnyru commented:
It took approximately the same time in total.
53m 18s

mathbunnyru commented on docs/maintaining/aarch64-runner.md:
````diff
@@ -0,0 +1,23 @@
+2. Run under `root`:
+
+   ```bash
+   /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/mathbunnyru/docker-stacks/asalikhov/new_build_system/aarch64-runner/setup.sh)"
````
I will change this before merging to master

mathbunnyru commented:
@trallard @consideRatio @parente @romainx @maresb this is done I believe 😂 

<img width="1416" alt="Снимок экрана 2022-07-04 в 21 53 37" src="https://user-images.githubusercontent.com/12270691/177200973-3b5eebc6-a270-4b53-b585-6dc71ff0215f.png">

Three runners are ready to use and working well.
Sharing data between them using artifacts works as well.
Docs are updated as well.
The description is also updated - please, read it to see, what's going on.

I will see why docs linkcheck fails later (even though I added an exception).

maresb commented:
This is really, really cool!!!

Now that it's up-and-running, I can see and understand how it works. The execution graph is beautiful!

I managed to quickly skim the entire PR contents. I didn't have time to do a properly thorough or critical review, but my impression is that this is really high-quality work. I feel like I learned a lot by reading through this. Excellent job!!! :+1:

mathbunnyru commented:
Thank you, @maresb!
I will also remove QEMU G_SLICE bug fix, which doesn't look good and makes aarch64 mamba command slower.

maresb commented:
Ah, right, since you're not using QEMU anymore. That's great!

But you could do that in a future PR. No reason to hold up this PR if it's passing.

mathbunnyru commented:
Yes, you're right, I can do it later.
Still, I want to receive at least one thorough review, because this is the biggest change I've ever done in this repo and it took half a year to actually come up with this.
Until I receive this review, I might change a thing or two, while waiting 😂

mathbunnyru commented:
@manics I forgot to mention you initially.
Please, if you have some time, review this PR :)

mathbunnyru commented:
I gave a bit more thought and I'm going to reconfigure my runners - I'm going to have two runners with 2 CPU each.
Some things are working faster, when there are several cores - extraction in mamba, tests running, (maybe) docker save.
So, we should see some speed improvement for building aarch64 images.

mathbunnyru commented:
I changed this option to require approval for all outside collaborators.
This is a simple measure to make self-hosted runners less vulnerable to arbitrary code execution.

<img width="783" alt="Снимок экрана 2022-07-05 в 11 12 00" src="https://user-images.githubusercontent.com/12270691/177270796-bc2679ae-d02e-4f63-91b7-6817cbdd0a95.png">

mathbunnyru commented:
It was with 3 runners with 1 CPU: [1h 38m 19s](https://github.com/jupyter/docker-stacks/actions/runs/2612384390/usage)
Now it's with 2 runners with 2 CPU: [1h 21m 29s](https://github.com/jupyter/docker-stacks/actions/runs/2614134590/usage)

The best we can get with this scheme is around 1h 11m.
To do this, we will have to have 6 self-hosted aarch64 runners.

This is actually something we might want to have, because in master branch pushing to registry will also take some time and we have only 2 runners, which means we're not doing it the best way we can.

benz0li commented:
@mathbunnyru Have you thought of running them on a Mac Mini (M1, 2020)?

I am hosting [GitLab Runner](https://gitlab.com/gitlab-org/gitlab-runner)s in a Debian VM on a Mac mini (M1, 2020) to build my `linux/arm64/v8` images.  
ℹ️ Even the first generation Mx machines outperform everything else out there.

mathbunnyru commented:
> @mathbunnyru Have you thought of running them on a Mac Mini (M1, 2020)?
> 
> I am hosting [GitLab Runner](https://gitlab.com/gitlab-org/gitlab-runner)s in a Debian VM on a Mac mini (M1, 2020) to build my `linux/arm64/v8` images. ℹ️ Even the first generation Mx machines outperform everything else out there.

I thought about running everything on Mac, but there is no free M1 Cloud Mac, as far as I know.
I might be wrong here.
You're using your personal Mac Mini, right?

Currently, 2 Oracle VMs with 2 CPUs each will be free for me.

benz0li commented:
> I thought about running aarch64 builds on M1 Mac, but there is no free M1 Cloud Mac, as far as I know.
> I might be wrong here.

No, you are not.

> You're using your personal Mac Mini, right?

Yes. It was a $900 investment.

mathbunnyru commented:
To be honest, I don't think it would be a good idea to use personal equipment in a very popular open-source project.
I mean, it's totally fine for personal projects and I would have a setup as you have for my personal projects (if I needed to).

But if something happens to the person who owns the hardware or he decides to stop providing it as a service or the hardware breaks, it would be quite difficult to make this project work again.
In my case, I added a complete guide to set up a new runner. If I decide to stop using Oracle cloud, then someone has to follow the instructions and it should work just fine. I did nothing to my VMs, except the things I documented. Moreover, I recreated everything from scratch after the last change, so it should be fine.
Another possibility is that Oracle stops providing these runners, but I hope it's not gonna happen and there are other alternatives (Amazon, for example, but not as good in a free tier).

mathbunnyru commented:
There were some reviews in the past, so I'm going to merge this when this passes.
Now, I have some time to fix with some bugs, if there will be any, in master branch.

mathbunnyru commented:
I was able to enable building `datascience-notebook` under aarch64.
And here comes the best part:
even in master with full `datascience-notebook` and pushing to DockerHub build time is [1h 23m 22s](https://github.com/jupyter/docker-stacks/actions/runs/2620522730/usage)

trallard commented:
Wow @mathbunnyru I have been away as I am stretched with work but kudos on this major refactoring. You've done an incredible work 🙂 thank you
